keras_cv/layers/preprocessing/repeated_augmentation.py

```diff
@@ -0,0 +1,113 @@
+# limitations under the License.
+import tensorflow as tf
+
+from keras_cv.layers.preprocessing.base_image_augmentation_layer import (
+    BaseImageAugmentationLayer,
+)
+
+
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class RepeatedAugmentation(BaseImageAugmentationLayer):
+    """RepeatedAugmentation augments each image in a batch multiple times.
+
+    This technique exists to emulate the behavior of stochastic gradient descent within
+    the context of mini-batch gradient descent.  When training large vision models,
+    choosing a large batch size can introduce too much noise into aggregated gradients
+    causing the overall batch's gradients to be less effective than gradients produced
+    using smaller gradients.  RepeatedAugmentation handles this by re-using the same
+    image multiple times within a batch creating correlated samples.
+
+    This layer increases your batch size by a factor of `len(augmenters)`.
+
+    Args:
+        augmenters: the augmenters to use to augment the image
+        shuffle: whether or not to shuffle the result.  Essential when using an
+            asynchronous distribution strategy such as ParameterServerStrategy.
+
+    Usage:
+
+    List of identical augmenters:
+    ```python
+    repeated_augment = cv_layers.RepeatedAugmentation(
+        augmenters=[cv_layers.RandAugment(value_range=(0, 255))] * 8
+    )
+    inputs = {
+        "images": tf.ones((8, 512, 512, 3)),
+        "labels": tf.ones((8,)),
+    }
+    outputs = repeated_augment(inputs)
+    # outputs now has a batch size of 64 because there are 8 augmenters
+    ```
+
+    List of distinct augmenters:
+    ```python
+    repeated_augment = cv_layers.RepeatedAugmentation(
+        augmenters=[
+            cv_layers.RandAugment(value_range=(0, 255)),
+            cv_layers.RandomFlip(),
+        ]
+    )
+    inputs = {
+        "images": tf.ones((8, 512, 512, 3)),
+        "labels": tf.ones((8,)),
+    }
+    outputs = repeated_augment(inputs)
+    ```
+
+    References:
+    - [DEIT implementaton](https://github.com/facebookresearch/deit/blob/ee8893c8063f6937fec7096e47ba324c206e22b9/samplers.py#L8)
+    - [Original publication](https://openaccess.thecvf.com/content_CVPR_2020/papers/Hoffer_Augment_Your_Batch_Improving_Generalization_Through_Instance_Repetition_CVPR_2020_paper.pdf)
+
+    """
+
+    def __init__(self, augmenters, shuffle=True, **kwargs):
+        super().__init__(**kwargs)
+        self.augmenters = augmenters
+        self.shuffle = shuffle
+
+    def _batch_augment(self, inputs):
+        if "bounding_boxes" in inputs:
+            raise ValueError(
+                "RepeatedAugmentation() does not yet support bounding box labels."
+            )
+
+        augmenter_outputs = [augmenter(inputs) for augmenter in self.augmenters]
+
+        outputs = {}
+        for k in inputs.keys():
+            outputs[k] = tf.concat([batch[k] for batch in augmenter_outputs], axis=0)
```

Comment:
Does this line concatenate images in batches of an epoch concatenated?
I dont get the concatenation. May I ask if you can add a comment for this part?

Reply:
In the original implementations of the repeated aug, there is a hyper parameter of number of repeats. Basically, how many times an augmented version of an image would exist inside a batch. I dont see that in this implementations.

More information:
Ignoring repeated aug, Augmentations are two parts:
primary augmentations that happen to all images (https://github.com/facebookresearch/deit/blob/ee8893c8063f6937fec7096e47ba324c206e22b9/augment.py#L98).
secondary augmentations that has a randomness inside. This will ensure that augmentations of an image are not the same (https://github.com/facebookresearch/deit/blob/ee8893c8063f6937fec7096e47ba324c206e22b9/augment.py#L111).

Reply:
An update from chat: we will meet on Tuesday to discuss more details of how RepeatedAugmentation is used and will figure out how that impacts the API.

Reply:
> An update from chat: we will meet on Tuesday to discuss more details of how RepeatedAugmentation is used and will figure out how that impacts the API.

As it seems they are using torch vision APJ are they going to apply the same transformation param within the batch? 
See the referecend doc and the Kornia option:
https://discuss.pytorch.org/t/random-batch-transforms/134173

Reply:
> As it seems they are using torch vision API are they going to apply the same transformation param within the batch?

Makes sense, but I do not have the engineering resources at this time to iron this out for this layer.  In the future I agree important - but I'll sort out the high priority KPL performances first.

Reply:
> but I'll sort out the high priority KPL performances first.

What do you mean?